kafka/record/legacy_records.py

```diff
@@ -0,0 +1,486 @@
+        pos += self.KEY_LENGTH
+        if key_size == -1:
+            key = None
+        else:
+            key = self._buffer[pos:pos + key_size].tobytes()
+            pos += key_size
+
+        value_size = struct.unpack_from(">i", self._buffer, pos)[0]
+        pos += self.VALUE_LENGTH
+        if value_size == -1:
+            value = None
+        else:
+            value = self._buffer[pos:pos + value_size].tobytes()
+        return key, value
+
+    def __iter__(self):
+        if self._magic == 1:
+            key_offset = self.KEY_OFFSET_V1
+        else:
+            key_offset = self.KEY_OFFSET_V0
+        timestamp_type = self.timestamp_type
+
+        if self.compression_type:
+            # In case we will call iter again
+            if not self._decompressed:
+                self._buffer = memoryview(self._decompress(key_offset))
+                self._decompressed = True
+
+            # If relative offset is used, we need to decompress the entire
+            # message first to compute the absolute offset.
+            headers = self._read_all_headers()
+            if self._magic > 0:
+                msg_header, _ = headers[-1]
+                absolute_base_offset = self._offset - msg_header[0]
+            else:
+                absolute_base_offset = -1
+
+            for header, msg_pos in headers:
+                offset, _, crc, _, attrs, timestamp = header
+                # There should only ever be a single layer of compression
+                assert not attrs & self.CODEC_MASK, (
+                    'MessageSet at offset %d appears double-compressed. This '
+                    'should not happen -- check your producers!' % offset)
+
+                # When magic value is greater than 0, the timestamp
+                # of a compressed message depends on the
+                # typestamp type of the wrapper message:
+                if timestamp_type == self.LOG_APPEND_TIME:
+                    timestamp = self._timestamp
+
+                if absolute_base_offset >= 0:
+                    offset += absolute_base_offset
+
+                key, value = self._read_key_value(msg_pos + key_offset)
+                yield LegacyRecord(
+                    offset, timestamp, timestamp_type,
+                    key, value, crc)
+        else:
+            key, value = self._read_key_value(key_offset)
+            yield LegacyRecord(
+                self._offset, self._timestamp, timestamp_type,
+                key, value, self._crc)
+
+
+class LegacyRecord(ABCRecord):
+
+    __slots__ = ("_offset", "_timestamp", "_timestamp_type", "_key", "_value",
+                 "_crc")
+
+    def __init__(self, offset, timestamp, timestamp_type, key, value, crc):
+        self._offset = offset
+        self._timestamp = timestamp
+        self._timestamp_type = timestamp_type
+        self._key = key
+        self._value = value
+        self._crc = crc
+
+    @property
+    def offset(self):
+        return self._offset
```

Comment:
I'm not familiar with this pattern of using @property to get private attributes. Is that primarily to "enforce" read-only behavior ?

Reply:
Yes, it's to enforce read only behaviour